### PR TITLE
Optional signing

### LINKED ISSUE #20: Allow test framework to refuse signing.
Would allow testing of cancelling initiated transfers as described in: https://github.com/aeternity/aeternity/pull/2806

### PR DESCRIPTION
**Pull request by AleksandarFilipov**
Re implement backchannel to work with the now optional signing.
- SocketConnector is now unaware of backchannel operations.

Fixed severe bug where retrurned socketchannel state wasn't saved.
Pause is now enforces, if messages arrvies during pause, they are thrown away.

Re added a few tests.

- Poi fetching is now optional when signing.
- Poi checks are available for arbitrary signing.

- Clean out code

fixes #20

### CODE REVIEW

File: apps/ae_socket_connector/lib/client_runner.ex [hunk 1]
```diff
@@ -20,7 +21,10 @@ defmodule ClientRunner do
   end
 
   defp log_callback(type, round, round_initiator, method, color) do
-    Logger.debug("received: #{inspect(type)}, #{inspect(round)}, #{inspect(round_initiator)}, #{inspect(method)}", color)
+    Logger.debug(
+      "received: #{inspect(type)}, #{inspect(round)}, #{inspect(round_initiator)}, #{inspect(method)}",
```
Comment (tolbrino) [inline apps/ae_socket_connector/lib/client_runner.ex:25]: In such cases you can also simply inspect a tuple with these elements, e.g. `#{inspect({type, round, round_initiator, method})}`

File: apps/ae_socket_connector/lib/client_runner.ex [hunk 2]
```diff
@@ -29,7 +33,7 @@ defmodule ClientRunner do
         Logger.debug(
           ":sign_approve, #{inspect(round)}, #{inspect(method)} extras: to_sign #{inspect(to_sign)} auto_approval: #{
             inspect(auto_approval)
```
Comment (tolbrino) [inline apps/ae_socket_connector/lib/client_runner.ex:35]: The formatting looks weird.

File: apps/ae_socket_connector/lib/client_runner.ex [hunk 3]
```diff
@@ -103,35 +107,85 @@ defmodule ClientRunner do
...
-      _ignore -> :ok
+  def process_sign_request({}, _to_sign, _pid_session_holder, _mode) do
+    Logger.debug("Empty request")
+  end
+
+  def process_sign_request(message, to_sign, pid_session_holder, %{sign: sign_info}) do
+    case elem(message, 0) do
+      :sign_approve ->
+        case sign_info do
+          {:default} ->
```
Comment (tolbrino) [inline apps/ae_socket_connector/lib/client_runner.ex:120]: Should be just `:default`, a tuple with a single element doesn't make much sense.

File: apps/ae_socket_connector/lib/client_runner.ex [hunk 4]
```diff
@@ -103,35 +107,85 @@ defmodule ClientRunner do
...
+            signed = SessionHolder.sign_message(pid_session_holder, to_sign)
+            fun = fn pid -> SocketConnector.send_signed_message(pid, elem(message, 2), signed) end
```
Comment (tolbrino) [inline apps/ae_socket_connector/lib/client_runner.ex]: Please don't use `elem/2` so much. Match the values up top somewhere and use them directly here. Same in the rest of the code.

Comment (tolbrino) [inline apps/ae_socket_connector/lib/client_runner.ex]: Simpler version:

```
fun = &(SocketConnector.send_signed_message(&1, elem(message, 2), signed))
```

File: apps/ae_socket_connector/lib/client_runner.ex [hunk 5]
```diff
@@ -103,35 +107,85 @@ defmodule ClientRunner do
...
+            SessionHolder.run_action(pid_session_holder, fun)
+
+          {:backchannel, pid_other_session_holder} ->
+            signed = SessionHolder.sign_message(pid_session_holder, to_sign)
+            signed2 = SessionHolder.sign_message(pid_other_session_holder, signed)
+            fun = fn pid -> SocketConnector.send_signed_message(pid, elem(message, 2), signed2) end
+            SessionHolder.run_action(pid_session_holder, fun)
+
+          {:check_poi} ->
+            fun = fn pid, from -> SocketConnector.get_poi(pid, from) end
```
Comment (tolbrino) [inline apps/ae_socket_connector/lib/client_runner.ex]:
```suggestion
            fun = &SocketConnector.get_poi/2
```

File: apps/ae_socket_connector/lib/client_runner.ex [hunk 6]
```diff
@@ -103,35 +107,85 @@ defmodule ClientRunner do
...
+    process_sign_request(message, to_sign, pid_session_holder)
+  end
+
+  def handle_cast({:end_pause}, state) do
+    GenServer.cast(self(), {:match_jobs, {}, nil})
+    {:noreply, %__MODULE__{state | paused: false}}
+  end
 
+  def handle_cast({:match_jobs, received_message, _to_sign}, %__MODULE__{paused: paused} = state)
+      when paused == true do
```
Comment (tolbrino) [inline apps/ae_socket_connector/lib/client_runner.ex]: Match `true` directly in the function head.

File: apps/ae_socket_connector/lib/client_runner.ex [hunk 7]
```diff
@@ -103,35 +107,85 @@ defmodule ClientRunner do
...
   end
 
-  # {:responder, :channels_info, 0, :transient, "channel_open"}
-  # def handle_cast({:match_jobs, message}, state)
-  def handle_cast({:match_jobs, received_message, to_sign}, state) do
-    process_sign_request(received_message, to_sign, state)
+  def process_sign_request(message, to_sign, pid_session_holder, _not_sign_request) do
+    process_sign_request(message, to_sign, pid_session_holder)
+  end
+
```
Comment (tolbrino) [inline apps/ae_socket_connector/lib/client_runner.ex:162]: General code layout: You mix public functions with gen_server callback implementations. I'd rather see some organization to keep these as groups separated with a comment before the groups.

**Review by tolbrino — APPROVED** [review]
Good improvement, I have lots of comments though :-) So keep that in mind.

File: apps/ae_socket_connector/lib/client_runner.ex (hunk 1 above)
Comment (AleksandarFilipov) [reply apps/ae_socket_connector/lib/client_runner.ex:25]: fixed

File: apps/ae_socket_connector/lib/client_runner.ex (hunk 4 above)
Comment (AleksandarFilipov) [reply apps/ae_socket_connector/lib/client_runner.ex]: fixed

File: apps/ae_socket_connector/lib/client_runner.ex (hunk 7 above)
Comment (AleksandarFilipov) [reply apps/ae_socket_connector/lib/client_runner.ex:162]: Agree will fix this later on.

**Pull request merged**